File: src/graph/BOJ9466.py
# https://www.acmicpc.net/problem/9466
# 텀프로젝트

import sys
sys.setrecursionlimit(100001)
input = sys.stdin.readline


def dfs(student):
    global count
    global visit
    global finish
    visit[student] = True
    nextStudent = choice[student]

    if not visit[nextStudent]:
        dfs(nextStudent)
    elif not finish[nextStudent]:
        s = nextStudent
        count += 1
        while(s != student):
            count += 1
            s = choice[s]

    finish[student] = True


for _ in range(int(input())):
    STUDENT_CNT = int(input())
    choice = [False] + list(map(int, input().split()))
    visit = [False] * (STUDENT_CNT+1)
    finish = [False] * (STUDENT_CNT+1)
    count = 0

    for student in range(1, STUDENT_CNT+1):
        if not visit[student]:
            dfs(student)
    print(STUDENT_CNT - count)

File: src/graph/BOJ9466.ts
export {};
// https://www.acmicpc.net/problem/9466
// 텀프로젝트

const inputStrings = require("fs")
  .readFileSync(process.platform === "linux" ? "/dev/stdin" : "./input.txt")
  .toString()
  .trim()
  .split("\n");

let answer = "";
let visit, finish: boolean[];
let count: number;

for (let i = 1; i < inputStrings.length; i++) {
  answer += `${solution(
    Number(inputStrings[i++]),
    inputStrings[i].split(" ").map(Number),
  )}\n`;
}

console.log(answer);

function solution(studentNumber: number, choice: number[]) {
  visit = new Array(studentNumber + 1);
  finish = new Array(studentNumber + 1);
  count = 0;
  for (let student = 1; student <= studentNumber; student++) {
    if (!visit[student]) dfs(student, choice);
  }
  return studentNumber - count;
}

function dfs(student: number, choice: number[]) {
  visit[student] = true;
  const nextStudent = choice[student - 1];
  if (!visit[nextStudent]) {
    dfs(nextStudent, choice);
  } else if (!finish[nextStudent]) {
    for (let i = nextStudent; i !== student; i = choice[i - 1]) {
      count++;
    }
    count++;
  }
  finish[student] = true;
}
